### ISSUE TITLE
handle_payment_update is not invoked when user returns from remote Moneris paypage

### ISSUE THREAD

**Issue by sharykin** [eventespresso/event-espresso-core#163]
> Before submitting your issue, please ensure that what you are reporting is specific to Event Espresso 4.

When submitting an issue, please follow this rough guideline for your report.

### Is this a bug report?
I'm not sure

### What branch did this occur in?
4.9.0.beta.002

### What version of WordPress is on the site this issue happened?
4.2.8

### Settings description

I have created Moneris gateway according to the instructions:

[EEG_Moneris.gateway.php.txt](https://github.com/eventespresso/event-espresso-core/files/282155/EEG_Moneris.gateway.php.txt)
[EE_PMT_Moneris.pm.php.txt](https://github.com/eventespresso/event-espresso-core/files/282158/EE_PMT_Moneris.pm.php.txt)

These two files renamed to php are located in payment_methods/Moneris folder of event espresso 4
They also contain logging.

I also added logging to
[EES_Espresso_Txn_Page.shortcode.php.txt](https://github.com/eventespresso/event-espresso-core/files/282167/EES_Espresso_Txn_Page.shortcode.php.txt)

The page that should be called by remote Moneris paypage is set in EE:

![jq_transaction_page](https://cloud.githubusercontent.com/assets/7074336/15543442/942efb08-229d-11e6-903e-d60a72282956.png)

On Moneris side the settings are:

![moneris_remote_settings](https://cloud.githubusercontent.com/assets/7074336/15543470/ad88b4c2-229d-11e6-9243-3af5582b040c.png)

### The issue
The system goes to the remote paypage. The payment goes through and the system goes to http://jq.cloudport.ca/transactions/.
When I check my log file it says only "set_redirection_info" which means nothing else has been invoked.
However it is not critical, I checked that the return data are exactly as they should be by setting the return data format on Moneris remote page to GET.
### The main question is:
- I set a transaction page with [ESPRESSO_TXN_PAGE] shortcode in it
- I set the remote Moneris paypage returning to my transaction page with POST data needed to update the payment

What else I have to do to get my handle_payment_update being invoked when this happens?

**Comment by joshfeck** [comment]
Hi there, 

It turns out there is some documentation that outlines how to set up your Gateway Class constructor for the handle_payment_update method:

https://github.com/eventespresso/event-espresso-core/blob/master/docs/L--Payment-Methods-and-Gateways/gateway-classes.md#handle_payment_update

In this case, it may be best if you can team up with a PHP developer to assist you with building out the custom gateway. We can recommend a few of the developers listed here:

http://eventespresso.com/developers/event-espresso-pros/

**Comment by sharykin** [comment]
Hi,

The documentation says that the default is when the user returns to espresso with IPN data from the Payment Gateway which is my case. There is nothing to set up in constructor as far as I see in the docs. But the method is not invoked no matter what I do. This is my question. What could be the reason causing it not being invoked?

Thank you for your proposition with developers but I can't hire anybody to solve my problem.

**Comment by joshfeck** [comment]
If it's sending the IPN data when the user returns from the Payment Gateway, then you do not use the Transactions page. 

> The method handle_payment_update() is called when an IPN is received, depending on whether your Gateway Class is set to receive IPN data in a separate request to the EE transactions page, *or* to receive IPN data when the user returns from the Payment Gateway.

italics added. So if the IPN data is received when the user returns, it's not a separate request to the EE transactions page.

**Comment by tn3rb** [comment]
@sharykin

to clarify, the Transactions page is for IPNs **ONLY** and does not support the actually user returning from the offsite gateway. If that is what you are attempting to do, then you need to send the user back to the Registration Checkout page. You can look at the PayPal Standard payment method to get an idea on how to do that.

**Comment by mnelson4** [comment]
yeah what's tricky about this integration, it seems, is that Moneris requires a static return url, whereas right now EE requires that it be dynamic (because we need to add the payment method's slug to the URL, and the primary registration's reg code, so that when users return from the offsite gateway which registration they're for and which gateway they're returning from). 
We have an experimental branch of EE that supports this: https://github.com/eventespresso/event-espresso-core/tree/FET-9334-gateway-data-router. That branch adds a new module named `EED_Gateway_Data_Router`, which has a method `get_static_return_url` (see `/modules/gateway_data_router/EED_Gateway_Data_Router.module.php`). This can be used to determine the right URL to set in Moneris for the "Approved URL". When the user returns to this URL, it should trigger your payment method's `handle_unclaimed_ipn` method where you'll need to use the POST/GET data to determine which payment to apply the transaction to. We'd like to spend more time on this but have competing priorities at the moment

**Comment by sharykin** [comment]
Thank you very much, I finally got it working. What was confusing me was the url that I was intrinsically thinking as about a static one. The one I set in Moneris config.
I took one url provided to set_redirection_info, removed transaction specific field spco_txn, and set this new url to Moneris. I transferred transaction id in an additional field which Moneris echos back. This way the loop worked.
Thank you guys again, I knew that something was wrong about my thinking, especially I didn't see method payment in my return url and that was suspicious.
Now everything looks right.
